replica/tablets.cc

```diff
@@ -474,10 +479,8 @@ stop_iteration tablet_sstable_set::for_each_sstable_set_until(const dht::partiti
 }
 
 future<stop_iteration> tablet_sstable_set::for_each_sstable_set_gently_until(const dht::partition_range& pr, std::function<future<stop_iteration>(lw_shared_ptr<sstables::sstable_set>)> func) const {
-    size_t candidate_start = pr.start() ? group_of(pr.start()->value().token()) : size_t(0);
-    size_t candidate_end = pr.end() ? group_of(pr.end()->value().token()) : (_sstable_sets.size() - 1);
-    for (auto i = candidate_start; i <= candidate_end; i++) {
-        if (const auto& set = _sstable_sets[i]) {
+    for (const auto& i : subrange(pr)) {
+        if (const auto& set = _sstable_sets.at(i)) {
```

Comment:
We shouldn't need to iterate over more than one tablet here. But let's leave that until multishard_reader dies.

Reply:
@denesb @bhalevy can this be called with a tablet-spanning range?

Reply:
storage_proxy::query_partition_key_range() needs to be taught about tablets:

1. hold erm
2. clamp range into first tablet that intersects it, remember if we changed the range
3. read page
4. has_more_pages |= we changed the range
5. return page

Reply:
we also have 

```
std::vector<sstables::shared_sstable> table::select_sstables(const dht::partition_range& range) const {
    return _sstables->select(range);
}
```

used by mutation dumper. we need to inspect every call in replica::table that is supplied with a token range, which can potentially cross tablets

if we come to think of it, tablet sstable set was introduced such that it can provide correct behavior while crossing tablets, isn't it good that we keep this capability even if we patch users to receive tablet treatment? It's orthogonal, and if it turns out one area or another can still cross tablets, we get correct behavior rather than aborting.

Reply:
There's never a reason to cross tablet replica bounds locally. If some shard owns tablet replica 37 and tablet replica 89, why would we want to read 89 just after 37?

Even SELECT mutation_fragments needs to jump among shards. So it should read shard 52 on a neighboring shard before 89.

Think of tablet replicas as independent nodes in a mini-rack. Would you want to read data from all nodes in a mini-rack? There's no relationship between them other than their physical location.

What we have here is baggage from multishard_reader, which in turn is due to tiny data ranges that made jumping between shards too expensive to do sequentially, so we did it concurrently. With tablets there's no such problem. It's more correct for SELECT mutation_fragment to chop ranges into tablets itself, or even better ask storage_proxy to do it.